### ISSUE TITLE
Documentation Enhancement: File Support

### ISSUE THREAD

**Issue by andybaer** [spring-projects/spring-integration#2930]
in the below 2 documents

- https://docs.spring.io/spring-integration/docs/5.1.0.RELEASE/reference/html/whats-new.html#x5.1-files
- https://docs.spring.io/spring-integration/docs/5.1.0.RELEASE/reference/html/files.html

there are many examples where the file-inbound-channel-adapter's property directory uses the file url prefix. according to my tests this does not work, it only works when specifiying the directory without any prefix or url formatting

`<int-file:inbound-channel-adapter id="filesIn"
    directory="file:${input.directory}" prevent-duplicates="true">
    <int-file:locker ref="customLocker"/>
</int-file:inbound-channel-adapter>`

**Comment by artembilan** [comment]
How does it look when you it doesn't work?
What an exception do you have?

Yes, I see it works well without `file:` when we have a `<context:property-placeholder/>`, but when we don't it doesn't resolve the property-placeholder at all.

Any possible way that we can reproduce it?
What is the minimal configuration can you share that we may confirm it is wrong?

Thanks

**Comment by garyrussell** [comment]
It works fine for me with or without a PPH.

Since the property is a `java.io.File`, Spring uses a `FileEditor`, which detects if the String is a URL and uses a `ResourceEditor` to convert the String to a Resource and then converts that to a `File`.

```java
	public void setAsText(String text) throws IllegalArgumentException {
		if (!StringUtils.hasText(text)) {
			setValue(null);
			return;
		}

		// Check whether we got an absolute file path without "file:" prefix.
		// For backwards compatibility, we'll consider those as straight file path.
		File file = null;
		if (!ResourceUtils.isUrl(text)) {
			file = new File(text);
			if (file.isAbsolute()) {
				setValue(file);
				return;
			}
		}

		// Proceed with standard resource location parsing.
		this.resourceEditor.setAsText(text);
		Resource resource = (Resource) this.resourceEditor.getValue();

		// If it's a URL or a path pointing to an existing resource, use it as-is.
		if (file == null || resource.exists()) {
			try {
				setValue(resource.getFile());
			}
			catch (IOException ex) {
				throw new IllegalArgumentException(
						"Could not retrieve file for " + resource + ": " + ex.getMessage());
			}
		}
		else {
			// Set a relative File reference and hope for the best.
			setValue(file);
		}
	}
```

**Comment by andybaer** [comment]
I configure "file:target" as directory and get the exception

`Source directory [file:target] does not exist.`

When I remove the prefix, it works. All other configuration like "autoCreateDirectory" etc does not matter. We are using integration versin 5.1.2

**Comment by artembilan** [comment]
That's not helpful. Sorry.

Any chances to have something we can play and reproduce from our side?
How does it work with the latest version - `5.1.5` ?

**Comment by andybaer** [comment]
I agree that this is most likely something on my side. It just can't be the permissions because it works with the same directory when I remove the prefix. I can report back with more details next week but feel free to close the issue. thanks

**Comment by garyrussell** [comment]
We can leave it open for a while, while you investigate. You might want to try setting s breakpoint in`FileEditor.setAsText()`.

**Comment by artembilan** [comment]
I think we don't need too much details. According to you description it is as simple as just `<int-file:inbound-channel-adapter id="filesIn" directory="file:${input.directory}" prevent-duplicates="true">` configuration.
So, I believe it should be easy for you to isolate an issue to a simple config and test-case for it. 
Or that can be just play Spring Boot project with that channel adapter configuration.
The point is that this `Source directory [file:target] does not exist.` happens just in the `start()`. So, there is no need in any other configuration.
Although, as far as I know, there is auto-configured properties placeholder in Spring Boot...

**Comment by Yanick-Salzmann** [comment]
I have created a zip with a minimal example that creates the same issue. 

[integration-dir-test.zip](https://github.com/spring-projects/spring-integration/files/3196628/integration-dir-test.zip)

```xml
	<int-file:inbound-channel-adapter
			id="fileInAdapter"
			channel="flow.in"
			directory="file:${output.directory}"
			filename-pattern="*"
			auto-create-directory="false"
			auto-startup="true">
		<int:poller fixed-delay="5000"/>
	</int-file:inbound-channel-adapter>
```

This results in:
`Caused by: java.lang.IllegalArgumentException: Source directory [file:C:\Users] does not exist.`

I am trying to debug how it loads the string and will give you another update in this issue.

**Comment by Yanick-Salzmann** [comment]
@garyrussell That breakpoint is not hit, and in `SourcePollingChannelAdapterFactoryBean#initializeAdapter` in the `FileReadingMessageSource` the directory is `file:C:\Users`.

**Comment by garyrussell** [comment]
Interesting. I see it too (on MacOS) with your sample.

Investigating...

**Comment by garyrussell** [comment]
The problem is because Boot injects a `ConversionService into the `BeanFactory` which pre-empts the property editor; the conversion service doesn't handle the conversion to a `Resource`.

I'll open an issue against Spring to see what the core team has to say.

**Comment by garyrussell** [comment]
Fixed in Boot 2.2.0.BUILD-SNAPSHOT https://github.com/spring-projects/spring-boot/issues/16931
